### PR TITLE
add item and user bias to loss, sgd, and trainining code

### PR DESCRIPTION
**Pull request by nickgreenquist**

### CODE REVIEW

File: matrix_factorization/tests/test_sgd.cu [hunk 1]
```diff
@@ -53,11 +53,38 @@ void test_sgd() {
     cudaMalloc(&errors_device, matrix->nonzeros * sizeof(float));
     cudaMemcpy(errors_device, errors, matrix->nonzeros * sizeof(float), cudaMemcpyHostToDevice);
 
+    // Create the bias arrays
+    float *user_bias = new float[rows];
```
Comment (dorukkilitcioglu) [inline matrix_factorization/tests/test_sgd.cu]: Since we're also training user bias, we should check if the updated biases are correct (see line 70).

File: matrix_factorization/tests/test_sgd.cu [hunk 2]
```diff
@@ -53,11 +53,38 @@ void test_sgd() {
...
+    for(int u = 0; u < rows; u++) {
+        user_bias[u] = 1.0;
+    }
+    float *user_bias_device;
+    cudaMalloc((void **) &user_bias_device, rows * sizeof(float));
+    cudaMemcpy(user_bias_device, user_bias, rows * sizeof(float), cudaMemcpyHostToDevice);
+
+    float *item_bias = new float[cols];
```
Comment (dorukkilitcioglu) [inline matrix_factorization/tests/test_sgd.cu]: Same as above

File: matrix_factorization/training.cu [hunk 3]
```diff
@@ -31,47 +30,90 @@ float *initialize_normal_array(int size) {
...
+        // Calculate initial error per each rating
+        calculate_loss_gpu(P_device, Q_device, n_factors, user_count, item_count, matrix->nonzeros, matrix,
+                           errors_device, user_bias_device, item_bias_device);
+
+        // Run single iteration of SGD
         sgd_update<<<dim_grid_sgd, dim_block>>>(matrix->indptr, matrix->indices, P_device->data, Q_device->data,
                                                 P_device_target->data, Q_device_target->data, n_factors, errors_device,
-                                                matrix->rows, matrix->cols, learning_rate);
+                                                user_count, item_count, learning_rate, user_bias_device, item_bias_device,
+                                                user_bias_target, item_bias_target, learning_rate, learning_rate);
```
Comment (dorukkilitcioglu) [inline matrix_factorization/training.cu]: lol, can we pass these in as params to the train function?

I guess we should have like a config struct that holds all these in, would make everything much easier to pass around, but that's for a different PR.

File: matrix_factorization/training.cu [hunk 4]
```diff
@@ -31,47 +30,90 @@ float *initialize_normal_array(int size) {
...
+        lastError = cudaGetLastError();
+        if(cudaSuccess != lastError) {
```
Comment (dorukkilitcioglu) [inline matrix_factorization/training.cu]: Can we check for error after the training loop? Feel like it might slow things down. Or maybe we can do it in the first iteration (to see if everything seems to be working) and one at the end?

File: matrix_factorization/training.cu [hunk 5]
```diff
@@ -31,47 +30,90 @@ float *initialize_normal_array(int size) {
...
+            printf("ERROR: %s\n", cudaGetErrorName(lastError));
+        }
+
+        // Calculate total loss to check for improving loss
         total_loss_kernel<<<dim_grid_loss, dim_block>>>(errors_device, losses_device, matrix->nonzeros, i);
+        lastError = cudaGetLastError();
+        if(cudaSuccess != lastError) {
```
Comment (dorukkilitcioglu) [inline matrix_factorization/training.cu]: See above

File: matrix_factorization/training.cu (hunk 3 above)
Comment (dorukkilitcioglu) [reply matrix_factorization/training.cu]: And by these, I mean the regularization terms.

File: matrix_factorization/loss.cu [hunk 6]
```diff
@@ -65,33 +61,21 @@ __global__ void total_loss_kernel(float *errors, float *losses, int n_errors, in
...
-    CudaDenseMatrix* output = new CudaDenseMatrix(1, 1, loss);
-
-    // Dimensions for kernel call
+void calculate_loss_gpu(CudaDenseMatrix* P_d, CudaDenseMatrix* Q_d, int factors, int user_count, int item_count, int num_ratings, 
+                        CudaCSRMatrix* matrix, float * error_d, float * user_bias,  float * item_bias) {
     int n_threads = 32;
     dim3 dimBlock(n_threads);
     dim3 dimGrid(user_count / n_threads + 1);
-    loss_kernel_user<<<dimGrid, dimBlock>>>(
+    cudaDeviceSynchronize();
```
[the remark this answers is not in the document]
Comment (nickgreenquist) [reply matrix_factorization/loss.cu]: no, probably something left over trying to make it work lol...will remove

File: matrix_factorization/tests/test_sgd.cu (hunk 1 above)
Comment (nickgreenquist) [reply matrix_factorization/tests/test_sgd.cu]: kk, ill find a way to test for that

File: matrix_factorization/training.cu (hunk 4 above)
Comment (nickgreenquist) [reply matrix_factorization/training.cu]: if you check after loop, you will only get an error from last kernel call, ie total_loss
